### PR TITLE
Add proper completion handling for challenge 

### PR DESCRIPTION
**Pull request by nikriek**

### CODE REVIEW

File: challenges/tests/test_models.py [hunk 1]
```diff
@@ -24,6 +25,27 @@ def test_challenge_metas(self):
...
+    def test_mark_as_completed(self):
+        for challenge_tuple in CHALLENGES:
+            challenge = get_challenge_factory(challenge_tuple[0])
+
+            self.assertFalse(challenge.mark_as_completed())
+            self.assertFalse(challenge.status == Challenge.COMPLETED)
+
+            def convenience_complete(challenge):
+                completion_fields = [field for field in challenge._meta.get_fields() if
+                                     field.name.endswith('_status')]
```
Comment (boeckhoff) [inline challenges/tests/test_models.py]: Looks like meta-programming? Is there a simpler way to do this?

File: challenges/models.py [hunk 2]
```diff
@@ -2,18 +2,19 @@
...
+    COMPLETED = "COMPLETED"
     ERROR = "ERROR"
 
     STATUS_CHOICES = (
         (NOT_STARTED, 'Not started'),
         (IN_PROGRESS, 'In progress'),
-        (DONE, 'Done'),
+        (COMPLETED, 'Completed'),
         (ERROR, 'Error'),
     )
```
Comment (boeckhoff) [inline challenges/models.py:19]: Why tuple of tuples instead of dict?

File: challenges/models.py [hunk 3]
```diff
@@ -46,10 +50,32 @@ def get_registered_steps_handlers(self):
...
-        step_func = self.get_registered_steps_handlers()[key]
+    def on_input(self, step_name, request):
+        step_func = self.get_registered_steps_handlers()[step_name]
         step_func(request)
 
+    def mark_as_completed(self, raise_exception=False):
+        """
+        Marks a challenge as completed if all fields ending with _status are also completed
+        """
+        fields = [field for field in self._meta.get_fields() if field.name.endswith('_status')]
```
Comment (boeckhoff) [inline challenges/models.py:67]: see above

File: challenges/tests/test_models.py (hunk 1 above)
Comment (nikriek) [reply challenges/tests/test_models.py]: Django actually uses a lot of meta programming. Not sure on how to map step function and status field

File: challenges/models.py (hunk 2 above)
Comment (nikriek) [reply challenges/models.py:19]: It's the Django way..

File: challenges/tests/test_models.py (hunk 1 above)
Comment (nikriek) [reply challenges/tests/test_models.py]: Any suggestions?

File: challenges/models.py (hunk 2 above)
Comment (nikriek) [reply challenges/models.py:19]: Tuples are immutable

**Comment by nikriek** [comment]
I would merge it soonish

**Pull request merged**